src/roaring64.c

```diff
@@ -1320,6 +1312,279 @@ void roaring64_bitmap_andnot_inplace(roaring64_bitmap_t *r1,
+    roaring_bitmap_t *bitmap32 = NULL;
+
+    // Iterate through buckets ordered by increasing keys.
+    while (it.value != NULL) {
+        uint32_t current_high32 = (uint32_t)(combine_key(it.key, 0) >> 32);
+        if (bitmap32 == NULL || prev_high32 != current_high32) {
+            if (bitmap32 != NULL) {
+                // Write as uint32 the most significant 32 bits of the bucket.
+                size += sizeof(prev_high32);
+
+                // Write the 32-bit Roaring bitmaps representing the least
+                // significant bits of a set of elements.
+                size += roaring_bitmap_portable_size_in_bytes(bitmap32);
+                roaring_bitmap_free_without_containers(bitmap32);
+            }
+
+            // Start a new 32-bit bitmap with the current high 32 bits.
+            art_iterator_t it2 = it;
+            uint32_t containers_with_high32 = 0;
+            while (it2.value != NULL && (uint32_t)(combine_key(it2.key, 0) >>
+                                                   32) == current_high32) {
+                containers_with_high32++;
+                art_iterator_next(&it2);
+            }
+            bitmap32 =
+                roaring_bitmap_create_with_capacity(containers_with_high32);
+
+            prev_high32 = current_high32;
+        }
+        leaf_t *leaf = (leaf_t *)it.value;
+        ra_append(&bitmap32->high_low_container,
+                  (uint16_t)(current_high32 >> 16), leaf->container,
+                  leaf->typecode);
+        art_iterator_next(&it);
+    }
+
+    if (bitmap32 != NULL) {
+        // Write as uint32 the most significant 32 bits of the bucket.
+        size += sizeof(prev_high32);
+
+        // Write the 32-bit Roaring bitmaps representing the least
+        // significant bits of a set of elements.
+        size += roaring_bitmap_portable_size_in_bytes(bitmap32);
+        roaring_bitmap_free_without_containers(bitmap32);
+    }
+
+    return size;
+}
+
+size_t roaring64_bitmap_portable_serialize(const roaring64_bitmap_t *r,
+                                           char *buf) {
+    // https://github.com/RoaringBitmap/RoaringFormatSpec#extension-for-64-bit-implementations
+    if (buf == NULL) {
+        return 0;
+    }
+    const char *initial_buf = buf;
+
+    // Write as uint64 the distinct number of "buckets", where a bucket is
+    // defined as the most significant 32 bits of an element.
+    uint64_t high32_count = count_high32(r);
+    memcpy(buf, &high32_count, sizeof(high32_count));
+    buf += sizeof(high32_count);
+
+    art_iterator_t it = art_init_iterator(&r->art, /*first=*/true);
+    bool first = true;
+    uint32_t prev_high32;
+    roaring_bitmap_t *bitmap32 = NULL;
+
+    // Iterate through buckets ordered by increasing keys.
+    while (it.value != NULL) {
+        uint64_t current_high48 = combine_key(it.key, 0);
+        uint32_t current_high32 = (uint32_t)(current_high48 >> 32);
+        if (first || prev_high32 != current_high32) {
```

Comment:
Sorry, same thing with `first` here, where I think it's clearer (to me at least) if we use `bitmap32 != NULL`, to me it makes more sense as "if we don't have a bitmap yet, or we need a new bitmap because we changed high bits, then get a new bitmap"

Reply:
Oops, forgot I did that twice :)